Adding support for grouped links

capajon opened the pull request:
To do:
- unit tests for item-group.  **UPDATE:** I don't expect to do this until Friday afternoon or so.  Hope it is okay to merge without.

Known Issues:
- ~~Critical: dropdown menu doesn't render when using "template iterations" (fixed in d2l-dropdown by Dave)~~
- High: Chrome only - opening menu causing nav area to "scroll" to second row of nav links that are normally cut-off.  Working in FF and IE. Weird.  **UPDATE:** Dave B discovered this is a common problem in Chrome with focus and overflow/hidden content.  We may need a custom event to be thrown to set the scrolltop to 0, or maybe we "live with it" until the "More" link is available (and the problem may go away).
- Low-Med: Firefox only - Arrow beside group text is 'pushed down' to a second line.  I noticed the profile area has a similar problem.  Flexbox is set to flex-wrap: nowrap (default).  Weird.  **UPDATE:** I'd like to follow up with a PR to address this separately.

@dbatiste 
@dlockhart

dlockhart commented on item-group.html:
```diff
@@ -0,0 +1,107 @@
+<dom-module id="d2l-navigation-item-group">
+	<style>
+		:host {
+			padding: 20px 20px 20px 0;
```
By default custom elements get `display: inline`, so you may want to put `display: inline-block` here so padding/margins/sizing work as you'd expect.

dlockhart commented on item-group.html:
```diff
@@ -0,0 +1,107 @@
+		}
+	</style>
+	<template>
+		<button	class="d2l-navigation-item-group-opener"	on-tap="_handleOpenerTap">
```
Missing space before `on-tap`

capajon replied:
Weird it doesn't show that way in my IDE.  Will address...

dlockhart commented on item-group.html:
```diff
@@ -0,0 +1,107 @@
+	</style>
+	<template>
+		<button	class="d2l-navigation-item-group-opener"	on-tap="_handleOpenerTap">
+			<span class="d2l-navigation-item-group-text vui-typography">{{groupText}}</span>
```
Not sure how I feel about using `vui-typography` here since we're just expecting that to be on the page. Plus, I'm not actually sure we want its style here... could you just do the same thing as the non-group items?

capajon replied:
This is a mistake and just leftover from something I was debugging/trying at some point.  Thanks for catching it!

dlockhart commented on item-group.html:
```diff
@@ -0,0 +1,107 @@
+			background: none;
+			border: 0;
+			color: inherit;
+			display: inline-flex;
```
I think you'll need the [browser prefixes](http://shouldiprefix.com/#flexbox) for any flex-box stuff to get it to work in IE/Safari/Firefox.

dlockhart commented on item-group.html:
```diff
@@ -0,0 +1,107 @@
+			text-decoration: underline;
+		}
+		.d2l-navigation-item-group-text {
+			margin-right: 0.3rem;
```
Do you need a RTL version for this?

capajon replied:
Yes :).  I did test RTL but didn't notice the margin missing.  Will fix.

capajon replied:
This is still to update.

dlockhart commented on item-group.html:
```diff
@@ -0,0 +1,107 @@
+			}
+			return '';
+		},
+		_computeLinks: function(item) {
```
Links can actually be quite complex and more than just `href`s -- they can open new windows (popups) and in some weird cases like "Logout" actually submit forms. I created an [item-behavior](https://github.com/Brightspace/d2l-navigation-ui/blob/master/item-behavior.html) behaviour to encapsulate all that logic of turning the HM data to the proper HREF and/or onclick. Hopefully you can reuse all that here? It also has logic to get the link text.

dbatiste replied:
@capajon : are you still going to want the ability to provide an href on the `d2l-menu-item`?

capajon replied:
While it is working okay as-is, I think it would be nice if menu item rendered a native link if provided a href (or vui-link :). The user could then get right-click functionality for optionally opening in a new tab, previewing a link etc.  

If that changes, we'd have to obviously update the template here to provide the href and also update this call to not pass true as the second parameter:
this.onItemLinkClick(this.links[index], true);

capajon commented:
Added three inline updates to the overview comment.  Also updated behaviors to be shared between single and group links, and a few oddities Dave and I encountered along the way.

Hoping it is close to merge worthy.

dlockhart commented on item-group.html:
```diff
@@ -0,0 +1,112 @@
+			if(item.entities.length > 0) {
+				item.entities.forEach(function(entity) {
+					if(entity.class.indexOf('link') > -1) {
+						links.push(entity);
```
You may be able to use Polymer's [list filtering](https://www.polymer-project.org/1.0/docs/devguide/templates#filtering-and-sorting-lists)?

capajon replied:
That's good to know about, actually.  Reading through it, it seems like how the index works might be a bit difficult to work with - it is needed in the listener "_handleLinkSelect" function.  So, I slightly prefer it as-is (unless you think there is a significant benefit to trying the Polymer filter).  But I'm also curious to try it and see if it works any better. :)

dlockhart replied:
Nope, I don't see a significant benefit, just thought it might make things easier.

dbatiste replied:
@capajon : perhaps you could use `item.entities.filter`? Then it would look like:
```
var links = item.entities.filter(function(entity) {
   return (entity.class.indexOf('link') > -1);
});
```

dlockhart commented:
Yep, looking good!

dbatiste commented on item-group.html:
```diff
@@ -0,0 +1,112 @@
+			display: -webkit-inline-box;
+			display: -moz-inline-box;
+			display: -ms-inline-flexbox;
+			display: -webkit-inline-flex;
```
At the risk of being a little naive with flexbox here: should there be 2 `-webkit-*` lines here?

capajon replied:
Yeah they generally appear in most "flex" mixins that I can find.  This seems to be the most popular order, though I think -webkit-inline-flex could safely move just below -webkit-inline-box.  There must be a storied history of webkit browser support behind it all. :)

dbatiste commented:
@capajon : 

> UPDATE: Dave B discovered this is a common problem in Chrome with focus and overflow/hidden content. We may need a custom event to be thrown to set the scrolltop to 0, or maybe we "live with it" until the "More" link is available (and the problem may go away).

Yeah, setting the `scrollTop = 0` on the element that has `overflow: hidden` was the only solution I could get to work.  A lot of other people have run into this issue.  Currently, `d2l-menu-item` fires an internal `_focus` event that you could listen for temporarily.  I can expose something more permanent if we feel it's needed, but my current thinking is that this solution should not be needed once nav item overflow is working as designed.

capajon commented:
@dbatiste Hmm, I think it worth setting the scrolltop for the first release.  I'll do that as part of a separate PR though.

Any objections to a merge?  I have the item above to follow up on, plus unit tests and the minor-ish firefox layout issue.

dbatiste commented:
Sounds good.

dlockhart commented:
Yep go for it
